app/upgrades/v16/upgrades.go

```diff
@@ -72,3 +81,38 @@ func CreateUpgradeHandler(
 		return mm.RunMigrations(ctx, configurator, vm)
 	}
 }
+
+// ConvertToNativeCoinExtensions converts all the registered ERC20 tokens of Cosmos native token pairs
+// back to the native representation and registers the (W)ERC20 precompiles for each token pair.
+func ConvertToNativeCoinExtensions(
+	ctx sdk.Context,
+	logger log.Logger,
+	accountKeeper authkeeper.AccountKeeper,
+	bankKeeper bankkeeper.Keeper,
+	erc20Keeper erc20keeper.Keeper,
+) error {
+	// NOTE (@fedekunze): first we must convert the all the registered tokens.
+	// If we do it the other way around, the conversion will fail since there won't
+	// be any contract code due to the selfdestruct.
+	if err := ConvertERC20Coins(ctx, logger, accountKeeper, bankKeeper, erc20Keeper); err != nil {
+		logger.Error("failed to convert native coins", "error", err.Error())
+		// TODO: return error here?
+		// return errorsmod.Wrap(err, "failed to convert native coins")
+	}
+
+	// Instantiate the (W)ERC20 Precompile for each registered IBC Coin
+
+	// IMPORTANT (@fedekunze): This logic needs to be included on EVERY UPGRADE
+	// from now on because the AvailablePrecompiles function does not have access
+	// to the state (in this case, the registered token pairs).
+	//
+	// FIXME: Do we want to run both if the convert coins failed? I think that could be dangerous if a coin
+	// was not fully converted and we overwrite the contract address with the Extension? Or maybe not?
+	if err := erc20Keeper.RegisterERC20Extensions(ctx); err != nil {
+		logger.Error("failed to register ERC-20 Extensions", "error", err.Error())
+		// TODO: return error here?
+		// return errorsmod.Wrap(err, "failed to convert native coins")
+	}
```

Comment:
Wondering if we should return an error here @fedekunze? If the conversion of the ERC20 coins fails we probably shouldn't go forward with replacing the contracts with the EVM extensions